### PR TITLE
Drop support for Python 2.7, 3.5, 3.6

### PR DESCRIPTION
**Pull request by dataflake**

### CODE REVIEW

File: .meta.toml [hunk 1]
```diff
@@ -2,14 +2,13 @@
 # https://github.com/zopefoundation/meta/tree/master/config/c-code
 [meta]
 template = "c-code"
-commit-id = "342271a70d886e753e5cc629e8a81b3cfab692ac"
+commit-id = "3c9d0741"
```
Comment (mgedmin) [inline .meta.toml:5]: Heh.

File: src/zope/proxy/__init__.py [hunk 2]
```diff
@@ -251,16 +251,15 @@ def __iter__(self):
...
+        return next(self._wrapped)
 
     def __next__(self):
         return self._wrapped.__next__()
 
-    # Python 2.7 won't let the C wrapper support __reversed__ :(
-    # def __reversed__(self):
-    #    return reversed(self._wrapped)
+    def __reversed__(self):
+        return reversed(self._wrapped)
```
Comment (mgedmin) [inline src/zope/proxy/__init__.py:262]: Note to self: check that the C wrapper supports it, and that this is mentioned in the changelog.

File: src/zope/proxy/__init__.py [hunk 3]
```diff
@@ -285,8 +284,8 @@ def __complex__(self):
     def __int__(self):
         return int(self._wrapped)
 
-    def __long__(self):  # pragma: no cover PY2
-        return long(self._wrapped)  # noqa: F821 undefined name
+    def __long__(self):  # Must stick around until zope.security is fixed
```
Comment (mgedmin) [inline src/zope/proxy/__init__.py:287]: Link to the zope.security github issue?

**Review by mgedmin — COMMENTED** [review]
LGTM except for one thing: I don't understand what's happening with `__reversed__`.

AFAICT the pure-python version of the proxy wrapper didn't support proxying `__reversed__`.  The C version didn't support it either on Python 2.7, if I can trust the deleted comment.  This PR added support for proxying `__reversed__` in the pure-python code but didn't touch the C code, nor the unit tests, nor  the changelog.

So, does the C wrapper support proxying `__reversed__`  do we have a unit test for that?  If we do, how come it didn't fail when the pure-python proxy didn't support it?  Was it always untested?

I remain somewhat confused.

File: src/zope/proxy/tests/test_proxy.py [hunk 4]
```diff
@@ -602,8 +542,8 @@ def test_binops(self):
                         z = eval(expr)
                         first = 0
                     else:
-                        self.assertEqual(eval(expr), z,
-                                         "x=%r; y=%r; expr=%r" % (x, y, expr))
+                        msg = "x={!r}; y={!r}; expr={!r}".format(x, y, expr)
```
[the remark this answers is not in the document]
Comment (icemac) [reply src/zope/proxy/tests/test_proxy.py:545]: As we now only support Python 3.7+ we can use f-strings wherever we want to.

[the remark this answers is not in the document]
Comment (dataflake) [reply src/zope/proxy/tests/test_proxy.py:545]: All this is done by `pyupgrade` which only creates f-strings in the simplest cases.

File: src/zope/proxy/__init__.py [hunk 5]
```diff
@@ -463,7 +454,7 @@ def __ipow__(self, other, modulus=None):
         return self
 
 
-AbstractPyProxyBase = _ProxyMetaclass(str('AbstractPyProxyBase'), (),
+AbstractPyProxyBase = _ProxyMetaclass('AbstractPyProxyBase', (),
```
[the remark this answers is not in the document]
Comment (dataflake) [reply src/zope/proxy/__init__.py:458]: `pyupgrade` did this. I don't know the motivation behind the original `str`.

File: src/zope/proxy/_zope_proxy_proxy.c [hunk 6]
```diff
@@ -890,16 +701,9 @@ ProxyType = {
...
-#if PY_MAJOR_VERSION < 3
     Py_TPFLAGS_DEFAULT |
     Py_TPFLAGS_HAVE_GC |
-    Py_TPFLAGS_CHECKTYPES |
     Py_TPFLAGS_BASETYPE,                    /* tp_flags */
-#else // Py_TPFLAGS_CHECKTYPES is always true in Python 3 and removed.
-    Py_TPFLAGS_DEFAULT |
-    Py_TPFLAGS_HAVE_GC |
-    Py_TPFLAGS_BASETYPE,                    /* tp_flags */
-#endif
```
[the remark this answers is not in the document]
Comment (dataflake) [reply src/zope/proxy/_zope_proxy_proxy.c:902]: I commented out the `__reversed__` definition in the Python module again. It was just a test to see what would happen. Researching this is a job for another day.

File: src/zope/proxy/__init__.py (hunk 2 above)
Comment (dataflake) [reply src/zope/proxy/__init__.py]: I commented the code again and added a note. This research should be done separately, right now I am just doing the mechanical job of cleaning out what I can.

File: src/zope/proxy/__init__.py (hunk 3 above)
Comment (dataflake) [reply src/zope/proxy/__init__.py]: I created the issue and added a link.

File: src/zope/proxy/_zope_proxy_proxy.c [hunk 7]
```diff
@@ -205,11 +187,6 @@ WrapperType_Lookup(PyTypeObject *type, PyObject *name)
...
 
         if (((PyTypeObject *)base) != &ProxyType) {
-#if PY_MAJOR_VERSION < 3
-            if (PyClass_Check(base))
-                dict = ((PyClassObject *)base)->cl_dict;
-            else
-#endif
             {
                 assert(PyType_Check(base));
                 dict = ((PyTypeObject *)base)->tp_dict;
```
[the remark this answers is not in the document]
Comment (dataflake) [reply src/zope/proxy/_zope_proxy_proxy.c]: Done

File: src/zope/proxy/tests/test_proxy.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dataflake) [reply src/zope/proxy/tests/test_proxy.py]: I changed the two places where `format` was used.

**Comment by dataflake** [comment]
Like I said, I re-commented `__reversed__`, that research can be done separately.

The `zope.proxy` code is closely tied to proxy code in `zope.security`. Both need a (separate) close look to remove all proxied names that are Python 2 only.

**Review by mgedmin — APPROVED** [review]
At first I was a bit torn when I saw you remove the `__bytes__` wrapper, but now I think that's right: the purpose of this PR is to drop support for old Python versions, not to fix missing wrappers for new Python versions.

(It's a bit strange that nobody has ever complained about `__bytes__` not being wrapped properly on Python 3 during all these years.)

**Pull request merged**